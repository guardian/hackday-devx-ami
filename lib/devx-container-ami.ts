import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import type { App } from 'aws-cdk-lib';
import {
	InstanceProfile,
	ManagedPolicy,
	Role,
	ServicePrincipal,
} from 'aws-cdk-lib/aws-iam';
import {
	CfnDistributionConfiguration,
	CfnImagePipeline,
	CfnImageRecipe,
	CfnInfrastructureConfiguration,
} from 'aws-cdk-lib/aws-imagebuilder';

export class DevxContainerAmi extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const recipe = new CfnImageRecipe(this, 'DevxContainerRecipe', {
			name: `devx-ami-${props.stage}-recipe`,
			parentImage:
				'arn:aws:imagebuilder:eu-west-1:aws:image/amazon-linux-2-arm64/2024.6.11',
			components: [
				{
					componentArn:
						'arn:aws:imagebuilder:eu-west-1:aws:component/amazon-corretto-17-jre/1.0.0/1',
				},
			],
			version: '1.0.0',
		});

		const role = new InstanceProfile(this, 'ImageBuilderProfie', {
			role: new Role(this, 'ImageBuilderRole', {
				assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
				managedPolicies: [
					ManagedPolicy.fromAwsManagedPolicyName(
						'EC2InstanceProfileForImageBuilder',
					),
				],
			}),
		});

		const config = new CfnInfrastructureConfiguration(
			this,
			'InfrastructureConfiguration',
			{
				name: `devx-ami-${props.stage}-config`,
				instanceProfileName: role.instanceProfileName,
			},
		);

		const distribution = new CfnDistributionConfiguration(
			this,
			'DevxContainerDistribution',
			{
				name: `devx-ami-${props.stage}-distribution`,
				distributions: [
					{
						region: 'eu-west-1',
						amiDistributionConfiguration: {},
					},
				],
			},
		);

		new CfnImagePipeline(this, 'DevxAmiPipeline', {
			name: `devx-ami-${props.stage}-pipeline`,
			description: 'Pipeline for creating the devx AMI',
			infrastructureConfigurationArn: config.attrArn,
			imageRecipeArn: recipe.attrArn,
			distributionConfigurationArn: distribution.attrArn,
		});
	}
}
