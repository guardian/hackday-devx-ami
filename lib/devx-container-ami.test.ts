import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { DevxContainerAmi } from "./devx-container-ami";

describe("The DevxContainerAmi stack", () => {
  it("matches the snapshot", () => {
    const app = new App();
    const stack = new DevxContainerAmi(app, "DevxContainerAmi", { stack: "playground", stage: "TEST" });
    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
