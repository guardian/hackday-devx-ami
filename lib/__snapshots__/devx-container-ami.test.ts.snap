// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The DevxContainerAmi stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [],
    "gu:cdk:version": "TEST",
  },
  "Resources": {
    "DevxAmiPipeline": {
      "Properties": {
        "Description": "Pipeline for creating the devx AMI",
        "DistributionConfigurationArn": {
          "Fn::GetAtt": [
            "DevxContainerDistribution",
            "Arn",
          ],
        },
        "ImageRecipeArn": {
          "Fn::GetAtt": [
            "DevxContainerRecipe",
            "Arn",
          ],
        },
        "InfrastructureConfigurationArn": {
          "Fn::GetAtt": [
            "InfrastructureConfiguration",
            "Arn",
          ],
        },
        "Name": "devx-ami-TEST-pipeline",
        "Tags": {
          "Stack": "playground",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/hackday-devx-ami",
        },
      },
      "Type": "AWS::ImageBuilder::ImagePipeline",
    },
    "DevxContainerDistribution": {
      "Properties": {
        "Distributions": [
          {
            "AmiDistributionConfiguration": {},
            "Region": "eu-west-1",
          },
        ],
        "Name": "devx-ami-TEST-distribution",
        "Tags": {
          "Stack": "playground",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/hackday-devx-ami",
        },
      },
      "Type": "AWS::ImageBuilder::DistributionConfiguration",
    },
    "DevxContainerRecipe": {
      "Properties": {
        "Components": [
          {
            "ComponentArn": "arn:aws:imagebuilder:eu-west-1:aws:component/amazon-corretto-17-jre/1.0.0/1",
          },
        ],
        "Name": "devx-ami-TEST-recipe",
        "ParentImage": "arn:aws:imagebuilder:eu-west-1:aws:image/amazon-linux-2-arm64/2024.6.11",
        "Tags": {
          "Stack": "playground",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/hackday-devx-ami",
        },
        "Version": "1.0.0",
      },
      "Type": "AWS::ImageBuilder::ImageRecipe",
    },
    "ImageBuilderProfie158DA968": {
      "Properties": {
        "Roles": [
          {
            "Ref": "ImageBuilderRole67171686",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "ImageBuilderRole67171686": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/EC2InstanceProfileForImageBuilder",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/hackday-devx-ami",
          },
          {
            "Key": "Stack",
            "Value": "playground",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "InfrastructureConfiguration": {
      "Properties": {
        "InstanceProfileName": {
          "Ref": "ImageBuilderProfie158DA968",
        },
        "Name": "devx-ami-TEST-config",
        "Tags": {
          "Stack": "playground",
          "Stage": "TEST",
          "gu:cdk:version": "TEST",
          "gu:repo": "guardian/hackday-devx-ami",
        },
      },
      "Type": "AWS::ImageBuilder::InfrastructureConfiguration",
    },
  },
}
`;
